stock_picking_propagate_scheduled_date/models/stock_move.py

```diff
@@ -30,4 +33,20 @@ def write(self, vals):
         # propagate date changes in the stock move chain
         if "date" in vals:
             self._propagate_date(vals.get("date"))
-        return super().write(vals)
+        return super(StockMove, self).write(vals)
+
+    def _register_hook(self):
+        # Patch `write` to initialize `date_propagated_ids` in the context, so that
+        # it's available on the whole stack chain, independent of the MRO.
+        res = super()._register_hook()
+
+        def make_write():
+            def write(self, vals, **kw):
+                if "date_propagation_ids" not in self.env.context:
+                    self = self.with_context(date_propagation_ids=set())
```

Comment:
I'm still wondering why initializing context with void value ?

Reply:
`_propagate_date` will take care of updating it:

```
        already_propagated_ids = self.env.context.get("date_propagation_ids")
        already_propagated_ids.update(self.ids)  # keep the same variable reference
```

We just need to make sure the variable exists, so that we can get its reference in `already_propagated_ids` by calling `self.env.context.get("date_propagation_ids")` instead of creating a new one

Reply:
And why not calling `already_propagated_ids = self.env.context.get("date_propagation_ids", set())`?

Reply:
That's essentially what was done before, but here's the issue: we initialize the `set` on the first `_propagate_date` call, and we're sure it remains in the context and is reused downstream. However, it may happen that some upstream code keeps on making changes to other records, modifying other records dates, and that these changes will trigger yet another `_propagate_date` call, where the context key is no longer there.

This is the case with the `mrp` and `mrp_subcontracting` core modules installed.

The MRO would look something like this:

```
>>> type(self.env['stock.move']).__mro__
(
    <class 'odoo.api.stock.move'>,
    ...
    <class 'odoo.addons.mrp_subcontracting.models.stock_move.StockMove'>,
    ...
    <class 'odoo.addons.stock_picking_propagate_scheduled_date.models.stock_move.StockMove'>,
    ...
    <class 'odoo.addons.mrp.models.stock_move.StockMove'>,
    ...
    <class 'odoo.addons.stock.models.stock_move.StockMove'>,
    <class 'odoo.models.Model'>,
    ...
    <class 'object'>
)
```

So, when the `write` method is executed with a new `date` in vals, the execution order will be the following (focusing only on the relevant modules)

```
mrp_subcontracting --> stock_picking_propagate_scheduled_date --> mrp
```

1) `mrp_subcontracting` will call `super()`: https://github.com/odoo/odoo/blob/3a680893a1d61a43eb8569ae9a48b7abbd6a86f5/addons/mrp_subcontracting/models/stock_move.py#L72

2) `stock_picking_propagate_scheduled_date` will initialize the context key, and do the propagation ✅ 

3) `mrp_subcontracting` will continue its flow after `super()`. Since this is upstream code, the context key isn't here ⚠️ . It will then propagate the move's `date` to the originating move's MO `date_planned_start` and `date_planned_finished`: https://github.com/odoo/odoo/blob/3a680893a1d61a43eb8569ae9a48b7abbd6a86f5/addons/mrp_subcontracting/models/stock_move.py#L77-L80

4) `mrp` will react to this change, and propagate them to its related `move_raw_ids` and `move_finished_ids`'s `date`: https://github.com/odoo/odoo/blob/5caec324326a197bb219a4d428b0072b543079b9/addons/mrp/models/mrp_production.py#L757-L760

5) This will get us back to the `stock.move` `write` method, without the context key being set. Depending on how these moves are related, the `date` change delta will be propagated again to the same moves that were propagated in the first run. 💀 


The solution is, then, to ensure the context key is in the context upstream in the code.

1) Before calling any module's override, the patched write will be executed first, ensuring the set is initialized in the context key

2) Then, the process continues exactly like before, but this time with the empty `set` already initialized in the context. So the already propagated ids are tracked throughout the whole call stack